### PR TITLE
Add a configurable "user tag" to SPDP

### PR DESCRIPTION
**Pull request by mitza-oci**
If the configuration specifies a non-zero tag, then that 32-bit value is added to the start of the SPDP participant announcement messages after the RTPS Header and its own Submessage Header.

### CODE REVIEW

File: dds/DCPS/RTPS/Spdp.cpp [hunk 1]
```diff
@@ -2747,11 +2760,30 @@ Spdp::SpdpTransport::write_i(WriteFlags flags)
   wbuff_.reset();
   DCPS::Serializer ser(&wbuff_, encoding_plain_native);
   DCPS::EncapsulationHeader encap(ser.encoding(), DCPS::MUTABLE);
-  if (!(ser << hdr_) || !(ser << data_) || !(ser << encap) || !(ser << plist)) {
+  if (!(ser << hdr_)) {
     if (DCPS::DCPS_debug_level > 0) {
```
Comment (iguessthislldo) [inline dds/DCPS/RTPS/Spdp.cpp:2764]:
```suggestion
    if (log_level >= LogLevel::Error) {
```

File: dds/DCPS/RTPS/Spdp.cpp [hunk 2]
```diff
@@ -2747,11 +2760,30 @@ Spdp::SpdpTransport::write_i(WriteFlags flags)
...
+        ACE_TEXT("(%P|%t) ERROR: Spdp::SpdpTransport::write_i: ")
+        ACE_TEXT("failed to serialize RTPS header for SPDP\n")));
+    }
+    return;
+  }
+  // The implementation-specific UserTagSubmessage is designed to directly
+  // follow the RTPS Message Header.  No other submessages should be added
+  // before it.  This enables filtering based on a fixed offset.
+  if (user_tag_.smHeader.submessageId && !(ser << user_tag_)) {
+    if (DCPS::DCPS_debug_level > 0) {
```
Comment (iguessthislldo) [inline dds/DCPS/RTPS/Spdp.cpp:2775]:
```suggestion
    if (log_level >= LogLevel::Error) {
```

File: dds/DCPS/RTPS/Spdp.cpp [hunk 3]
```diff
@@ -2747,11 +2760,30 @@ Spdp::SpdpTransport::write_i(WriteFlags flags)
...
+      ACE_ERROR((LM_ERROR,
+        ACE_TEXT("(%P|%t) ERROR: Spdp::SpdpTransport::write_i: ")
+        ACE_TEXT("failed to serialize user tag for SPDP\n")));
+    }
+    return;
+  }
+  if (!(ser << data_) || !(ser << encap) || !(ser << plist)) {
+    if (DCPS::DCPS_debug_level > 0) {
```
Comment (iguessthislldo) [inline dds/DCPS/RTPS/Spdp.cpp:2783]:
```suggestion
    if (log_level >= LogLevel::Error) {
```

**Review by iguessthislldo — CHANGES_REQUESTED** [review]

**Review by jrw972 — APPROVED** [review]

**Review by iguessthislldo — APPROVED** [review]

File: docs/devguide/run_time_configuration.rst [hunk 4]
```diff
@@ -1321,6 +1321,14 @@ Those properties, along with options specific to OpenDDS's RTPS Discovery implem
 
      - 1 (enabled)
 
+   * - ``SpdpUserTag=i``
+
+     - Add the OpenDDS-specific UserTag RTPS submessage to the start of SPDP messages.
+       If i is 0 (the default), the submessage is not added.
+       Otherwise this submessage's contents is the 4-byte integer i.
```
[the remark this answers is not in the document]
Comment (mitza-oci) [reply docs/devguide/run_time_configuration.rst:1328]: I was intending it to be more general about "an integer" than the C++ data type, but maybe it's better to be specific.

**Pull request merged**